File: mobile_client/lib/voice/audio_queue.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'package:audioplayers/audioplayers.dart';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';

import 'pcm_chunker.dart';
import 'pcm_stream_player.dart';

/// Sequential playback of assistant audio clips. The voice backend hands
/// us either MP3 segments (quality mode, and the realtime MP3 fallback)
/// or raw 24 kHz PCM-S16LE frames (realtime). PCM is wrapped in a WAV
/// header; clips play back-to-back.
///
/// We write each clip to a temp file and play it via [DeviceFileSource]
/// rather than [BytesSource] — on iOS, BytesSource silently fails to
/// decode (play() "completes" instantly and you hear nothing).
///
/// Realtime PCM does NOT wait for a whole segment: [appendPcm] streams it
/// through a [PcmChunker], so the first ~160 ms plays while the rest of the
/// sentence is still arriving (plan 1.7). The whole-buffer [enqueuePcm] and the
/// temp-file MP3 path are unchanged and remain the fallback.
///
/// [onIdle] fires when the queue drains (so the controller can return to
/// listening/idle); [onAmplitude] drives the orb during playback.
class AudioQueue {
  AudioQueue({
    this.onIdle,
    this.onAmplitude,
    this.onPlaybackStart,
    this.onClipStart,
    this.onPosition,
  }) {
    // Force playback onto the speaker and let it coexist with an active
    // recording session. In realtime the mic keeps the iOS session in
    // `.playAndRecord`, where playback otherwise routes to the (quiet)
    // earpiece.
    final ctx = AudioContext(
      iOS: AudioContextIOS(
        category: AVAudioSessionCategory.playAndRecord,
        options: const {
          AVAudioSessionOptions.defaultToSpeaker,
          AVAudioSessionOptions.mixWithOthers,
          AVAudioSessionOptions.allowBluetoothA2DP,
        },
      ),
      android: const AudioContextAndroid(
        contentType: AndroidContentType.speech,
        usageType: AndroidUsageType.assistant,
        audioFocus: AndroidAudioFocus.gainTransientMayDuck,
      ),
    );
    AudioPlayer.global.setAudioContext(ctx);
    _player.setAudioContext(ctx);
    _player.onPlayerComplete.listen((_) => _advance());
    unawaited(_native.probe());
    // Live playback position — drives the karaoke word highlight. Only the
    // current clip's tag is reported so the controller can map position →
    // word within the right segment.
    _player.onPositionChanged.listen((p) {
      // Positions are reported SEGMENT-relative: a segment split into several
      // streamed chunks must still advance one continuous word schedule, so we
      // add the audio of this segment already played by earlier chunks.
      if (_currentTag != null) {
        onPosition?.call(
            _currentTag, p + Duration(milliseconds: _currentClipBaseMs));
      }
    });
  }

  final void Function()? onIdle;
  final void Function(double amp)? onAmplitude;
  final void Function()? onPlaybackStart;

  /// Fired when a tagged clip begins, with the clip's playback [duration]
  /// (exact for PCM, queried for MP3). [tag] is the caller's clip id.
  final void Function(int? tag, Duration duration)? onClipStart;

  /// Fired ~5×/sec with the current clip's playback [position].
  final void Function(int? tag, Duration position)? onPosition;

  final AudioPlayer _player = AudioPlayer();
  final List<_Clip> _queue = [];
  bool _playing = false;
  bool _stopped = false;
  int _seq = 0;
  int? _currentTag; // tag of the clip currently playing (for position events)
  int _currentClipBaseMs = 0; // ms of this tag's audio played by earlier chunks
  PcmChunker? _chunker; // streaming cutter for the realtime PCM path (fallback)

  // ── Native gapless stream (plan 1.7, preferred) ─────────────────────────
  // One continuous render stream per reply instead of a temp file + play()
  // per slice — the per-slice player swap was the audible "cuts out every
  // second" stutter. Karaoke timing is derived from bytes fed + wall clock.
  final PcmStreamPlayer _native = PcmStreamPlayer();
  Future<void> _nativeChain = Future.value(); // keeps feed() order
  bool _nativeActive = false; // stream open and audio queued/playing
  bool _nativeEnded = false; // audio_end seen; go idle once playback catches up
  int? _nativeTag; // segment currently being fed
  int _nativeFedMs = 0; // total audio handed to the native player this stream
  int _nativeSegBaseMs = 0; // fed ms when the current segment began
  DateTime? _nativeStart; // wall clock at first feed
  Timer? _nativeTick;
  bool _nativePending = false; // a feed is queued on the chain but not yet run
  DateTime? _nativeLastFeed; // wall clock of the last feed (stall safety)
  static const int _kNativeTickMs = 100; // karaoke position cadence
  static const int _kNativeIdleGraceMs = 120; // wait for the last buffer to drain
  // Safety net: if audio_end never arrives (dropped frame, server error) go
  // idle once everything fed has played and nothing new came for this long.
  static const int _kNativeStallIdleMs = 1500;

  /// Enqueue an MP3 clip (raw bytes). [tag] (a segment id) lets the caller
  /// sync a word highlight to this clip's playback.
  void enqueueMp3(Uint8List bytes, {int? tag}) {
    if (bytes.isEmpty) return;
    debugPrint('[audio] enqueue MP3 ${bytes.length} bytes');
    _enqueue(_Clip(bytes, 'mp3', tag: tag));
  }

  /// Enqueue a raw PCM-S16LE clip at [sampleRate] (mono), wrapped in a
  /// WAV container. PCM duration is exact (computed from byte count).
  void enqueuePcm(Uint8List pcm, {int sampleRate = 24000, int? tag}) {
    if (pcm.isEmpty) return;
    debugPrint('[audio] enqueue PCM ${pcm.length} bytes @ ${sampleRate}Hz');
    final durMs = pcm.lengthInBytes ~/ 2 * 1000 ~/ sampleRate;
    _enqueue(_Clip(_wrapWav(pcm, sampleRate), 'wav', tag: tag, durationMs: durMs));
  }

  /// Stream realtime PCM for segment [tag] as it arrives. Plays the leading
  /// slice as soon as there's enough of it instead of waiting for the segment
  /// to finish (plan 1.7). Call [endPcmSegment] when `audio_end` lands.
  void appendPcm(Uint8List pcm, {int sampleRate = 24000, int? tag}) {
    if (_stopped || pcm.isEmpty) return;
    if (_native.available && _queue.isEmpty && !(_playing && !_nativeActive)) {
      _appendNative(pcm, sampleRate, tag);
      return;
    }
    final chunker = _chunkerFor(sampleRate);
    for (final chunk in chunker.add(pcm, tag: tag)) {
      _enqueueChunk(chunk);
    }
  }

  /// The segment is complete — play whatever is left of it.
  void endPcmSegment({int? tag}) {
    if (_stopped) return;
    if (_nativeActive || (_native.available && _nativePending)) {
      // Ordered behind the feeds already queued on the chain: a short reply's
      // audio_end can land before the first feed() has even opened the stream,
      // and marking "ended" ahead of that feed would be overwritten by it.
      _nativeChain = _nativeChain.then((_) {
        if (_nativeActive) _nativeEnded = true;
      });
      return;
    }
    final chunker = _chunker;
    if (chunker == null) return;
    for (final chunk in chunker.flush(tag: tag)) {
      _enqueueChunk(chunk);
    }
  }

  /// True while a segment is mid-stream (some audio buffered but not emitted).
  bool get hasPendingPcm =>
      _nativeActive || (_chunker?.bufferedBytes ?? 0) > 0;

  void _appendNative(Uint8List pcm, int sampleRate, int? tag) {
    final ms = pcm.lengthInBytes ~/ 2 * 1000 ~/ sampleRate;
    _nativePending = true;
    _nativeChain = _nativeChain.then((_) async {
      _nativePending = false;
      if (_stopped) return;
      if (!_nativeActive) {
        if (!await _native.start(sampleRate)) {
          // Native side refused — hand this chunk to the fallback path.
          for (final chunk in _chunkerFor(sampleRate).add(pcm, tag: tag)) {
            _enqueueChunk(chunk);
          }
          return;
        }
        _nativeActive = true;
        _nativeEnded = false;
        _playing = true;
        _nativeFedMs = 0;
        _nativeSegBaseMs = 0;
        _nativeStart = null;
        _nativeTag = tag;
      }
      if (tag != _nativeTag) {
        // Next sentence in the same stream: positions restart at its base.
        _nativeSegBaseMs = _nativeFedMs;
        _nativeTag = tag;
        _currentClipBaseMs = 0;
      }
      _nativeEnded = false;
      await _native.feed(pcm);
      _nativeFedMs += ms;
      _nativeLastFeed = DateTime.now();
      if (_nativeStart == null) {
        _nativeStart = DateTime.now();
        _currentTag = tag;
        onPlaybackStart?.call();
        onAmplitude?.call(0.6);
        _startNativeTick();
      }
      _currentTag = tag;
      // Running total for this segment so the karaoke schedule stretches as
      // more of the sentence arrives (a repeat tag = schedule correction).
      onClipStart?.call(tag, Duration(milliseconds: _nativeFedMs - _nativeSegBaseMs));
    });
  }

  void _startNativeTick() {
    _nativeTick?.cancel();
    _nativeTick = Timer.periodic(const Duration(milliseconds: _kNativeTickMs), (_) {
      final start = _nativeStart;
      if (!_nativeActive || start == null) return;
      final elapsed = DateTime.now().difference(start).inMilliseconds;
      final played = elapsed.clamp(0, _nativeFedMs);
      final tag = _nativeTag;
      if (tag != null) {
        onPosition?.call(tag, Duration(milliseconds: played - _nativeSegBaseMs));
      }
      if (_nativeEnded && elapsed >= _nativeFedMs + _kNativeIdleGraceMs) {
        _finishNative(fireIdle: true);
        return;
      }
      final last = _nativeLastFeed;
      if (!_nativePending &&
          last != null &&
          elapsed >= _nativeFedMs + _kNativeStallIdleMs &&
          DateTime.now().difference(last).inMilliseconds >= _kNativeStallIdleMs) {
        debugPrint('[audio] native stream: no audio_end — idling after stall');
        _finishNative(fireIdle: true);
      }
    });
  }

  void _finishNative({required bool fireIdle}) {
    _nativeTick?.cancel();
    _nativeTick = null;
    _nativeActive = false;
    _nativeEnded = false;
    _nativeStart = null;
    _nativeLastFeed = null;
    _nativeTag = null;
    _currentTag = null;
    _playing = false;
    unawaited(_native.stop());
    onAmplitude?.call(0);
    if (fireIdle) onIdle?.call();
  }

  PcmChunker _chunkerFor(int sampleRate) {
    final existing = _chunker;
    if (existing != null && existing.sampleRate == sampleRate) return existing;
    return _chunker = PcmChunker(sampleRate: sampleRate);
  }

  void _enqueueChunk(PcmChunk chunk) {
    if (chunk.bytes.isEmpty) return;
    _enqueue(_Clip(
      _wrapWav(chunk.bytes, chunk.sampleRate),
      'wav',
      tag: chunk.tag,
      durationMs: chunk.durationMs,
      // Announce the segment's running total, not this slice's length, so the
      // karaoke schedule stretches as more of the sentence arrives.
      segmentMs: chunk.segmentMsSoFar,
      baseMs: chunk.segmentMsSoFar - chunk.durationMs,
    ));
  }

  void _enqueue(_Clip clip) {
    if (_stopped) return;
    _queue.add(clip);
    if (!_playing) _advance();
  }

  Future<void> _advance() async {
    if (_stopped) return;
    if (_queue.isEmpty) {
      if (_playing) {
        _playing = false;
        onAmplitude?.call(0);
        onIdle?.call();
      }
      return;
    }
    _playing = true;
    final clip = _queue.removeAt(0);
    try {
      final dir = await getTemporaryDirectory();
      final path = '${dir.path}/jc_voice_${_seq++}.${clip.ext}';
      final file = File(path);
      await file.writeAsBytes(clip.bytes, flush: true);
      // Drop position events while we swap clips so a trailing event from the
      // outgoing clip isn't attributed to the incoming one.
      _currentTag = null;
      await _player.stop();
      await _player.play(DeviceFileSource(path));
      _currentTag = clip.tag;
      _currentClipBaseMs = clip.baseMs;
      onPlaybackStart?.call();
      onAmplitude?.call(0.6); // coarse "speaking" pulse for the orb
      debugPrint('[audio] playing $path (${clip.bytes.length}b)');
      // Announce the clip + its duration so the controller can schedule the
      // word highlight. PCM is exact; MP3 we ask the player (may be 0 until
      // it loads — the controller falls back to a words×rate estimate).
      if (clip.tag != null) {
        if (clip.durationMs != null) {
          // PCM: exact duration, schedule the karaoke immediately. For a
          // streamed segment we report the running TOTAL so far — the
          // controller treats a repeat tag as a schedule correction, not a
          // restart, so the highlight keeps advancing across chunk seams.
          onClipStart?.call(clip.tag,
              Duration(milliseconds: clip.segmentMs ?? clip.durationMs!));
        } else {
          // MP3: getDuration() right after play() is unreliable on iOS (it returns
          // null or a stale/wrong value), which gives the karaoke a wrong word
          // schedule — the highlight sticks on the first word, then jumps to the
          // end when the next clip starts. Start on an estimate now (so _curSeg is
          // set and words can advance), then correct the schedule when the REAL
          // decoded duration arrives via onDurationChanged.
          onClipStart?.call(clip.tag, Duration.zero);
          _correctDuration(clip.tag);
        }
      }
    } catch (e, st) {
      debugPrint('[audio] play() FAILED: $e\n$st');
      _advance(); // skip a bad clip rather than wedging the queue
    }
  }

  /// MP3 duration isn't reliably known at play() time, so we re-fire [onClipStart]
  /// with the REAL decoded duration once it lands — from [AudioPlayer.onDurationChanged]
  /// (a getDuration() fallback covers the case it already fired). Applied at most
  /// once per clip, and only while [tag] is still the current clip.
  void _correctDuration(int? tag) {
    StreamSubscription<Duration>? sub;
    Timer? fallback;
    var applied = false;
    void apply(Duration? d) {
      if (applied) return;
      applied = true;
      sub?.cancel();
      fallback?.cancel();
      if (_currentTag == tag && d != null && d.inMilliseconds > 0) {
        onClipStart?.call(tag, d);
      }
    }

    sub = _player.onDurationChanged.listen(apply);
    fallback = Timer(const Duration(milliseconds: 350), () async {
      Duration? d;
      try {
        d = await _player.getDuration();
      } catch (_) {}
      apply(d);
    });
  }

  /// Drop anything queued and stop playback immediately (barge-in / new
  /// turn). Does NOT fire [onIdle].
  Future<void> stop() async {
    _queue.clear();
    _playing = false;
    _currentTag = null;
    _currentClipBaseMs = 0;
    // Barge-in/new turn: drop half-assembled audio too, or the tail of the
    // interrupted sentence would play after the user has already moved on.
    _chunker?.reset();
    if (_nativeActive) {
      await _native.flush();
      _finishNative(fireIdle: false);
    }
    try {
      await _player.stop();
    } catch (_) {}
    onAmplitude?.call(0);
  }

  bool get isBusy => _playing || _queue.isNotEmpty || _nativeActive;

  Future<void> dispose() async {
    _stopped = true;
    _nativeTick?.cancel();
    unawaited(_native.stop());
    _queue.clear();
    _chunker?.reset();
    await _player.dispose();
  }

  /// Minimal 44-byte PCM WAV header + samples.
  static Uint8List _wrapWav(Uint8List pcm, int sampleRate) {
    const channels = 1;
    const bitsPerSample = 16;
    final byteRate = sampleRate * channels * bitsPerSample ~/ 8;
    const blockAlign = channels * bitsPerSample ~/ 8;
    final dataLen = pcm.length;
    final out = BytesBuilder();
    void str(String s) => out.add(s.codeUnits);
    void u32(int v) => out.add([
          v & 0xFF,
          (v >> 8) & 0xFF,
          (v >> 16) & 0xFF,
          (v >> 24) & 0xFF,
        ]);
    void u16(int v) => out.add([v & 0xFF, (v >> 8) & 0xFF]);

    str('RIFF');
    u32(36 + dataLen);
    str('WAVE');
    str('fmt ');
    u32(16); // PCM fmt chunk size
    u16(1); // audio format = PCM
    u16(channels);
    u32(sampleRate);
    u32(byteRate);
    u16(blockAlign);
    u16(bitsPerSample);
    str('data');
    u32(dataLen);
    out.add(pcm);
    return out.toBytes();
  }
}

class _Clip {
  _Clip(this.bytes, this.ext, {this.tag, this.durationMs, this.segmentMs, this.baseMs = 0});
  final Uint8List bytes;
  final String ext;
  final int? tag; // caller's segment id, for karaoke highlight sync
  final int? durationMs; // exact for PCM; null for MP3 (queried at play)

  /// Total audio of [tag] emitted so far, including this clip. Only set on the
  /// streaming PCM path; null means "this clip IS the whole segment".
  final int? segmentMs;

  /// Audio of [tag] played by EARLIER clips — added to every position report
  /// so the word highlight is continuous across a split segment.
  final int baseMs;
}

File: ios_app/JarvisCopilot/Copilot/Voice/AudioQueue.swift
import Foundation

/// Sequential playback of assistant audio clips. Port of `voice/audio_queue.dart`.
///
/// The voice backend hands us either MP3 segments (quality mode, and the realtime
/// MP3 fallback) or raw 24 kHz PCM-S16LE frames (realtime). Two paths:
///
///  • **Gapless stream** (preferred, plan 1.7) — realtime PCM does NOT wait for a
///    whole segment: `appendPcm` feeds one continuous render stream, so the first
///    ~160 ms plays while the rest of the sentence is still arriving. Karaoke
///    timing is derived from bytes fed + wall clock.
///  • **Clip queue** (fallback + every MP3) — clips play back-to-back; PCM is cut
///    by a `PcmChunker` and wrapped in a WAV header.
///
/// `onIdle` fires when the queue drains (so the store can return to
/// listening/idle); `onAmplitude` drives the orb during playback.
@MainActor
final class AudioQueue {

    // MARK: - Callbacks

    var onIdle: (() -> Void)?
    var onAmplitude: ((Double) -> Void)?
    var onPlaybackStart: (() -> Void)?
    /// A tagged clip began, with the clip's playback duration in ms (exact for
    /// PCM, the running segment total on the streaming path). `tag` is the
    /// caller's karaoke segment id.
    var onClipStart: ((Int?, Int) -> Void)?
    /// The current clip's playback position in ms.
    var onPosition: ((Int?, Int) -> Void)?

    // MARK: - Tuning

    /// Karaoke position cadence on the streaming path.
    static let nativeTickMs = 100
    /// Wait for the last buffer to drain before going idle.
    static let nativeIdleGraceMs = 120
    /// Safety net: if `audio_end` never arrives (dropped frame, server error) go
    /// idle once everything fed has played and nothing new came for this long.
    static let nativeStallIdleMs = 1500

    // MARK: - Dependencies

    private let output: AudioOutput
    private let clock: VoiceClock

    // MARK: - Clip queue state

    private struct Clip {
        let bytes: Data
        let ext: String
        /// Caller's segment id, for karaoke highlight sync.
        let tag: Int?
        /// Exact for PCM; nil for MP3 (the decoder tells us at play time).
        let durationMs: Int?
        /// Total audio of `tag` emitted so far, including this clip. Only set on
        /// the streaming-fallback path; nil means "this clip IS the whole segment".
        let segmentMs: Int?
        /// Audio of `tag` played by EARLIER clips — added to every position
        /// report so the word highlight is continuous across a split segment.
        let baseMs: Int
    }

    private var queue: [Clip] = []
    private var playing = false
    private var stopped = false
    /// Tag of the clip currently playing (for position events).
    private var currentTag: Int?
    /// Ms of this tag's audio played by earlier chunks.
    private var currentClipBaseMs = 0
    /// An MP3 clip waiting for its real decoded duration.
    ///
    /// The player can report the duration from INSIDE `play()` — before we have
    /// announced the estimate — so the correction is buffered and applied strictly
    /// after it, or the estimate would overwrite the good schedule.
    private struct DurationCorrection {
        let tag: Int
        var estimateAnnounced = false
        var durationMs: Int?
    }
    private var correction: DurationCorrection?
    /// Streaming cutter for the realtime PCM fallback.
    private var chunker: PcmChunker?

    // MARK: - Gapless stream state

    private var nativeActive = false
    /// `audio_end` seen; go idle once playback catches up.
    private var nativeEnded = false
    /// Segment currently being fed.
    private var nativeTag: Int?
    /// Total audio handed to the player this stream.
    private var nativeFedMs = 0
    /// Fed ms when the current segment began.
    private var nativeSegBaseMs = 0
    private var nativeStart: Date?
    private var nativeTick: VoiceTimerToken?
    /// A feed is queued on the chain but has not yet run.
    private var nativePending = false
    private var nativeLastFeed: Date?

    // MARK: - Serialised async work
    // One chain for both paths so `startStream`/`feed`/`play` can never
    // interleave — out-of-order audio is the one failure you would hear.

    private var chain: Task<Void, Never>?
    private var chainSeq = 0
    /// Bumped by `stop`/`dispose` so work already queued becomes a no-op.
    private var epoch = 0

    init(output: AudioOutput, clock: VoiceClock) {
        self.output = output
        self.clock = clock

        output.onClipComplete = { [weak self] in self?.advance() }
        output.onClipPosition = { [weak self] pos in
            guard let self, let tag = self.currentTag else { return }
            // Positions are reported SEGMENT-relative: a segment split into
            // several streamed chunks must still advance one continuous word
            // schedule, so we add the audio already played by earlier chunks.
            self.onPosition?(tag, Int(pos * 1000) + self.currentClipBaseMs)
        }
        output.onClipDuration = { [weak self] dur in
            guard let self, var pending = self.correction, dur > 0 else { return }
            // MP3 duration isn't known at play() time, so we re-fire onClipStart
            // with the REAL duration once it lands. The store treats a repeat tag
            // as a schedule correction, not a restart.
            let ms = Int(dur * 1000)
            if pending.estimateAnnounced {
                self.correction = nil
                self.onClipStart?(pending.tag, ms)
            } else {
                pending.durationMs = ms
                self.correction = pending
            }
        }
    }

    // MARK: - Public state

    var isBusy: Bool { playing || !queue.isEmpty || nativeActive }

    /// True while a segment is mid-stream (some audio buffered but not emitted).
    var hasPendingPcm: Bool { nativeActive || (chunker?.bufferedBytes ?? 0) > 0 }

    // MARK: - Enqueue

    /// Enqueue an MP3 clip (raw bytes). `tag` (a segment id) lets the caller sync
    /// a word highlight to this clip's playback.
    func enqueueMp3(_ bytes: Data, tag: Int? = nil) {
        guard !bytes.isEmpty else { return }
        enqueue(Clip(bytes: bytes, ext: "mp3", tag: tag,
                     durationMs: nil, segmentMs: nil, baseMs: 0))
    }

    /// Enqueue a whole raw PCM-S16LE clip at `sampleRate` (mono), wrapped in a
    /// WAV container. PCM duration is exact (computed from the byte count).
    func enqueuePcm(_ pcm: Data, sampleRate: Int = 24000, tag: Int? = nil) {
        guard !pcm.isEmpty else { return }
        let durMs = (pcm.count / 2) * 1000 / sampleRate
        enqueue(Clip(bytes: Self.wrapWav(pcm, sampleRate: sampleRate), ext: "wav",
                     tag: tag, durationMs: durMs, segmentMs: nil, baseMs: 0))
    }

    /// Stream realtime PCM for segment `tag` as it arrives. Plays the leading
    /// slice as soon as there's enough of it instead of waiting for the segment
    /// to finish. Call `endPcmSegment` when `audio_end` lands.
    func appendPcm(_ pcm: Data, sampleRate: Int = 24000, tag: Int? = nil) {
        guard !stopped, !pcm.isEmpty else { return }
        if output.isStreamAvailable, queue.isEmpty, !(playing && !nativeActive) {
            appendNative(pcm, sampleRate: sampleRate, tag: tag)
            return
        }
        for chunk in chunkerFor(sampleRate).add(pcm, tag: tag) { enqueueChunk(chunk) }
    }

    /// The segment is complete — play whatever is left of it.
    func endPcmSegment(tag: Int? = nil) {
        guard !stopped else { return }
        if nativeActive || (output.isStreamAvailable && nativePending) {
            // Ordered behind the feeds already queued: a short reply's audio_end
            // can land before the first feed has even opened the stream, and
            // marking "ended" ahead of that feed would be overwritten by it.
            let ep = epoch
            serial { [weak self] in
                guard let self, ep == self.epoch, self.nativeActive else { return }
                self.nativeEnded = true
            }
            return
        }
        guard let chunker else { return }
        for chunk in chunker.flush(tag: tag) { enqueueChunk(chunk) }
    }

    /// Drop anything queued and stop playback immediately (barge-in / new turn).
    /// Does NOT fire `onIdle`.
    func stop() async {
        epoch += 1
        // The per-frame chain is only ever as long as the audio still to play;
        // dropping it here stops a barge-in paying for the whole interrupted
        // sentence, and stops the last task being retained until the next one.
        chain?.cancel()
        chain = nil
        queue.removeAll()
        playing = false
        currentTag = nil
        currentClipBaseMs = 0
        correction = nil
        // Barge-in/new turn: drop half-assembled audio too, or the tail of the
        // interrupted sentence would play after the user has moved on.
        chunker?.reset()
        if nativeActive {
            await output.flushStream()
            finishNative(fireIdle: false)
        }
        await output.stopClip()
        onAmplitude?(0)
    }

    func dispose() async {
        stopped = true
        epoch += 1
        chain?.cancel()
        chain = nil
        nativeTick?.cancel()
        nativeTick = nil
        queue.removeAll()
        chunker?.reset()
        await output.stopStream()
        await output.stopClip()
    }

    /// Tests only: wait until every queued async step has run (a step may queue
    /// more, so we drain until the chain stops growing).
    func settle() async {
        var last = -1
        for _ in 0..<64 {
            if chainSeq == last { return }
            last = chainSeq
            await chain?.value
        }
    }

    // MARK: - Gapless stream path

    private func appendNative(_ pcm: Data, sampleRate: Int, tag: Int?) {
        let ms = (pcm.count / 2) * 1000 / sampleRate
        nativePending = true
        let ep = epoch
        serial { [weak self] in
            guard let self else { return }
            self.nativePending = false
            guard !self.stopped, ep == self.epoch else { return }
            if !self.nativeActive {
                guard await self.output.startStream(sampleRate: sampleRate) else {
                    // The player refused — hand this chunk to the fallback path.
                    for chunk in self.chunkerFor(sampleRate).add(pcm, tag: tag) {
                        self.enqueueChunk(chunk)
                    }
                    return
                }
                self.nativeActive = true
                self.nativeEnded = false
                self.playing = true
                self.nativeFedMs = 0
                self.nativeSegBaseMs = 0
                self.nativeStart = nil
                self.nativeTag = tag
            }
            if tag != self.nativeTag {
                // Next sentence in the same stream: positions restart at its base.
                self.nativeSegBaseMs = self.nativeFedMs
                self.nativeTag = tag
                self.currentClipBaseMs = 0
            }
            self.nativeEnded = false
            await self.output.feed(pcm)
            guard ep == self.epoch else { return }
            self.nativeFedMs += ms
            self.nativeLastFeed = self.clock.now
            if self.nativeStart == nil {
                self.nativeStart = self.clock.now
                self.currentTag = tag
                self.onPlaybackStart?()
                self.onAmplitude?(0.6) // coarse "speaking" pulse for the orb
                self.armNativeTick()
            }
            self.currentTag = tag
            // Running total for this segment so the karaoke schedule stretches as
            // more of the sentence arrives (a repeat tag = schedule correction).
            self.onClipStart?(tag, self.nativeFedMs - self.nativeSegBaseMs)
        }
    }

    private func armNativeTick() {
        nativeTick?.cancel()
        nativeTick = clock.schedule(after: Self.nativeTickMs) { [weak self] in
            self?.nativeTickFired()
        }
    }

    private func nativeTickFired() {
        nativeTick = nil
        guard nativeActive, let start = nativeStart else { return }
        let elapsed = Int(clock.now.timeIntervalSince(start) * 1000)
        let played = min(max(elapsed, 0), nativeFedMs)
        if let tag = nativeTag { onPosition?(tag, played - nativeSegBaseMs) }
        if nativeEnded, elapsed >= nativeFedMs + Self.nativeIdleGraceMs {
            finishNative(fireIdle: true)
            return
        }
        if !nativePending, let last = nativeLastFeed,
           elapsed >= nativeFedMs + Self.nativeStallIdleMs,
           Int(clock.now.timeIntervalSince(last) * 1000) >= Self.nativeStallIdleMs {
            // No audio_end — idle after the stall rather than hanging in "speaking".
            finishNative(fireIdle: true)
            return
        }
        armNativeTick()
    }

    private func finishNative(fireIdle: Bool) {
        nativeTick?.cancel()
        nativeTick = nil
        nativeActive = false
        nativeEnded = false
        nativeStart = nil
        nativeLastFeed = nil
        nativeTag = nil
        currentTag = nil
        playing = false
        let ep = epoch
        serial { [weak self] in
            guard let self, ep == self.epoch else { return }
            await self.output.stopStream()
        }
        onAmplitude?(0)
        if fireIdle { onIdle?() }
    }

    // MARK: - Clip queue path

    private func chunkerFor(_ sampleRate: Int) -> PcmChunker {
        if let existing = chunker, existing.sampleRate == sampleRate { return existing }
        let fresh = PcmChunker(sampleRate: sampleRate)
        chunker = fresh
        return fresh
    }

    private func enqueueChunk(_ chunk: PcmChunk) {
        guard !chunk.bytes.isEmpty else { return }
        enqueue(Clip(bytes: Self.wrapWav(chunk.bytes, sampleRate: chunk.sampleRate),
                     ext: "wav",
                     tag: chunk.tag,
                     durationMs: chunk.durationMs,
                     // Announce the segment's running total, not this slice's
                     // length, so the karaoke schedule stretches as more of the
                     // sentence arrives.
                     segmentMs: chunk.segmentMsSoFar,
                     baseMs: chunk.segmentMsSoFar - chunk.durationMs))
    }

    private func enqueue(_ clip: Clip) {
        guard !stopped else { return }
        queue.append(clip)
        if !playing { advance() }
    }

    private func advance() {
        guard !stopped else { return }
        if queue.isEmpty {
            if playing {
                playing = false
                onAmplitude?(0)
                onIdle?()
            }
            return
        }
        playing = true
        let clip = queue.removeFirst()
        let ep = epoch
        serial { [weak self] in
            guard let self, ep == self.epoch else { return }
            // Drop position events while we swap clips so a trailing event from
            // the outgoing clip isn't attributed to the incoming one.
            self.currentTag = nil
            // Armed BEFORE play(), because play() may report the duration itself.
            self.correction = (clip.durationMs == nil ? clip.tag : nil)
                .map { DurationCorrection(tag: $0) }
            await self.output.stopClip()
            guard ep == self.epoch else { return }
            guard await self.output.play(clip.bytes, fileExtension: clip.ext) else {
                self.advance() // skip a bad clip rather than wedging the queue
                return
            }
            guard ep == self.epoch else { return }
            self.currentTag = clip.tag
            self.currentClipBaseMs = clip.baseMs
            self.onPlaybackStart?()
            self.onAmplitude?(0.6)
            guard clip.tag != nil else { return }
            if let dur = clip.durationMs {
                // PCM: exact duration, schedule the karaoke immediately. For a
                // streamed segment we report the running TOTAL so far.
                self.onClipStart?(clip.tag, clip.segmentMs ?? dur)
            } else {
                // MP3: start on an estimate now (so the segment is current and
                // words can advance), then correct with the real duration.
                self.onClipStart?(clip.tag, 0)
                self.correction?.estimateAnnounced = true
                if let pending = self.correction, let ms = pending.durationMs {
                    self.correction = nil
                    self.onClipStart?(pending.tag, ms)
                }
            }
        }
    }

    private func serial(_ work: @escaping @MainActor () async -> Void) {
        let previous = chain
        chainSeq += 1
        chain = Task { @MainActor in
            await previous?.value
            await work()
        }
    }

    // MARK: - WAV

    /// Minimal 44-byte PCM WAV header + samples.
    static func wrapWav(_ pcm: Data, sampleRate: Int) -> Data {
        let channels = 1, bitsPerSample = 16
        let byteRate = sampleRate * channels * bitsPerSample / 8
        let blockAlign = channels * bitsPerSample / 8
        var out = Data(capacity: 44 + pcm.count)
        func str(_ s: String) { out.append(contentsOf: Array(s.utf8)) }
        func u32(_ v: Int) {
            let u = UInt32(truncatingIfNeeded: v)
            out.append(contentsOf: [UInt8(u & 0xFF), UInt8((u >> 8) & 0xFF),
                                    UInt8((u >> 16) & 0xFF), UInt8((u >> 24) & 0xFF)])
        }
        func u16(_ v: Int) {
            let u = UInt16(truncatingIfNeeded: v)
            out.append(contentsOf: [UInt8(u & 0xFF), UInt8((u >> 8) & 0xFF)])
        }
        str("RIFF"); u32(36 + pcm.count); str("WAVE")
        str("fmt "); u32(16); u16(1); u16(channels)
        u32(sampleRate); u32(byteRate); u16(blockAlign); u16(bitsPerSample)
        str("data"); u32(pcm.count)
        out.append(pcm)
        return out
    }
}
